baseplate/context/thrift.py

```diff
@@ -52,25 +65,48 @@ def _enumerate_service_methods(client):
 class PooledClientProxy(object):
     """A proxy which acts like a thrift client but uses a connection pool."""
 
-    def __init__(self, client_cls, pool, root_span, namespace):
+    def __init__(self, client_cls, pool, root_span, namespace, retry_policy=None):
         self.client_cls = client_cls
         self.pool = pool
         self.root_span = root_span
         self.namespace = namespace
+        self.retry_policy = retry_policy or RetryPolicy.new(attempts=1)
 
         for name in _enumerate_service_methods(client_cls):
             setattr(self, name, functools.partial(
                 self._call_thrift_method, name))
 
+    @contextlib.contextmanager
+    def retrying(self, **policy):
+        yield PooledClientProxy(
+            self.client_cls,
+            self.pool,
+            self.root_span,
+            self.namespace,
+            retry_policy=RetryPolicy.new(**policy),
+        )
+
     def _call_thrift_method(self, name, *args, **kwargs):
         trace_name = "{}.{}".format(self.namespace, name)
-
-        with self.root_span.make_child(trace_name) as span:
-            with self.pool.connection() as prot:
-                prot.trans.set_header("Trace", str(span.trace_id))
-                prot.trans.set_header("Parent", str(span.parent_id))
-                prot.trans.set_header("Span", str(span.id))
-
-                client = self.client_cls(prot)
-                method = getattr(client, name)
-                return method(*args, **kwargs)
+        last_error = None
+
+        for _ in self.retry_policy:
+            try:
+                with self.root_span.make_child(trace_name) as span:
+                    with self.pool.connection() as prot:
+                        prot.trans.set_header("Trace", str(span.trace_id))
+                        prot.trans.set_header("Parent", str(span.parent_id))
+                        prot.trans.set_header("Span", str(span.id))
+
+                        client = self.client_cls(prot)
+                        method = getattr(client, name)
+                        return method(*args, **kwargs)
```

Comment:
The `with self.pool.connection() as prot` can raise `TTransportException` [here](https://github.com/reddit/baseplate/blob/retries/baseplate/thrift_pool.py#L72) if other requests are using all the pooled connections, or [here](https://github.com/reddit/baseplate/blob/retries/baseplate/thrift_pool.py#L100) if the pool exhausts its retry budget attempting to connect. The `method` can then raise `TTransportException` if the connection times out or otherwise breaks during its service call.

The only unsafe thing in here is `method` which might be a non-idempotent service method on the remote service. The goal of the `with` statement syntax for overriding the policy is that it forces the user to think about whether or not it's safe to retry the operation. We'd need to extend thrift somehow to indicate idempotency if we wanted this to be checked automatically.

I dithered on whether the span should live outside or inside the retry logic, but ended on it being within. The reason I chose this is that each `make_child` call generates a new span ID for the child span and this means that instead of sending the same span id to the remote service N times, we send a different one for each retry. This is all rather theoretical right now as we still don't have Zipkin running, but it seems to me that'd be a more useful representation to the remote service since it has no reason to care that the N calls it just got were part of the same retry.